### DISCUSSION TITLE
Understanding memory usage

### DISCUSSION THREAD

**Discussion by AndrewHannigan** [networkit/networkit#715] [Q&A]
I am working with a 3BN+ edge graph so I've had a keen eye on memory consumption.  To benchmark memory consumption I've created a bipartite graph with 10MN nodes on each side of the graph where all nodes have degree 1.  I am seeing a memory consumption of roughly 0.5KB per edge.  This is higher than I expected from the following back of the envelope math:

Node array of length 20MN = 20MN * size(node)
Total space occupied by all adjacency lists = 10MN * size(node)
Total space occupied by all weight lists = 10MN * size(node)

So with a size(node) of 8 bytes, that would be roughly 40MN * 8bytes (0.3GB) for 10MN edges. That would be 0.03kb per edge.  What I'm seeing instead for a graph of this scale is a total memory consumption of 4.5GB, or 0.45kb per edge.  

My question is whether 0.5kb per edge is expected for memory consumption, and if so what am I missing in the above math?  Is this related to the way that adjacency lists are pre-allocated?

Below is the script I am using to benchmark memory usage:
```
# Usage
# /path/to/this/script.py N_NODES DEGREE

import numpy as np
import networkit as nk
from memory_profiler import profile

EDGELIST_PATH = 'example_graph.edgelist'

def write_edgelist(n=10_000, d=2):
    arrays = []
    for i in range(d):
        x = np.zeros([n, 3], dtype=int)
        x[:, 0] = range(0, n)
        x[:, 1] = range(n, n * 2)
        x[:, 2] = np.random.randint(100, size=n)

        np.random.shuffle(x[:, 1])
        arrays.append(x)

    final = np.concatenate(arrays)

    np.savetxt(EDGELIST_PATH, final, delimiter='\t', fmt='%i')

@profile
def read_edgelist():
    edgelist_reader = nk.graphio.EdgeListReader(separator='\t',
                                                firstNode=0,
                                                continuous=False,
                                                directed=True
                                                )
    g = edgelist_reader.read(EDGELIST_PATH)


if __name__ == '__main__':
    import sys
    N = int(sys.argv[1])
    D = int(sys.argv[2])

    print(f"Running benchmark for {N=} and {D=}")

    write_edgelist(n=N, d=D)
    read_edgelist()
```

**Comment by fabratu** [comment]
Thats a good question. There are likely maintainers more involved in this matter, but I can have a start. 

These are the basic data-structures in a graph causing memory consumption:

```
    //!< exists[v] is true if node v has not been removed from the graph
    std::vector<bool> exists;

    //!< only used for directed graphs, inEdges[v] contains all nodes u that
    //!< have an edge (u, v)
    std::vector<std::vector<node>> inEdges;
    //!< (outgoing) edges, for each edge (u, v) v is saved in outEdges[u] and
    //!< for undirected also u in outEdges[v]
    std::vector<std::vector<node>> outEdges;

    //!< only used for directed graphs, same schema as inEdges
    std::vector<std::vector<edgeweight>> inEdgeWeights;
    //!< same schema (and same order!) as outEdges
    std::vector<std::vector<edgeweight>> outEdgeWeights;

    //!< only used for directed graphs, same schema as inEdges
    std::vector<std::vector<edgeid>> inEdgeIds;
    //!< same schema (and same order!) as outEdges
    std::vector<std::vector<edgeid>> outEdgeIds;
```

So besides the already mentioned structures in your math - an additional ID-vector is created. Also you read in a directed graph which means an inEdges, outEdges (+ weights) are created. For these vectors the initial length is the 20MN * size(node) without entries. If I am not mistaken the final size of these is more like (for the directed case):

Edges: size(inEdges init) + size(outEdges init) + 2 * size(existing Edges since in/out) = 20MN * size(node) + 20MN * size(node) + 20MN * size(node) = 60MN * size(node)
Weights: the same as above = 60MN * size(node)
EdgeIds: the same as above = 60MN * size(node)

Neglecting the exists-vector and plus the nodes from your math (20MN * size(node)) - the final result is in the range of:  200MN * size(node) = 1.5 GB

This is already somewhat closer to your result.

There is also likely some overhead created "by" the reader itself. The final number of nodes/edges is not known when starting reading. Thats why the vectors are not pre-allocated, but filled incrementally. The final size of the vector may therefore use more memory.

So to (not) answer your question: Not sure, what the final memory consumption per node/edge is in reality. You could try to use the NetworKitBinaryReader (with writing the file beforehand to disc and clearing memory afterwards). This is more optimized in terms of speed and memory usage.

**Comment by avdgrinten** [comment]
There is some overhead because memory allocators usually round up allocations (e.g., of edge lists) to discrete size classes.

There are plans to improve that in the next few months, but I cannot give you an ETA.

**Comment by AndrewHannigan** [comment]
@avdgrinten could you point me towards the main point in the code that specifies the discrete size classes? Is it something like [2, 4, 8, 16, etc]?

**Comment by AndrewHannigan** [comment]
@fabratu do you have experience with writing to nk’s binary format from outside of nk? The issue that I’m currently experiencing is that I’m exporting the raw graph data from bigquery, so for the initial graph load into memory I think I’m restricted to using a text based format like EdgeList. Once I’ve created the nk binary version of the graph though I’ve seen lightning fast read speeds. Seeing roughly the same memory consumption though when reading from binary, which is probably the discrete size classes @avdgrinten mentioned.

Seems that there is a balance here between speed of the read and the size of the resulting in-memory graph.  The more discrete size classes we introduce the slower the read time but the more lean the resulting graph.  At the moment graphs with low avg degree seem to take the hardest hit in terms of space efficiency, so perhaps introducing smaller size classes would be helpful? Down the road having a knob that let’s you tune the balance between read speed and in-memory graph size would be a very neat feature.

**Comment by michitux** [comment]
I assume with discrete size classes @avdgrinten indeed means the doubling of the memory size of the vector's memory allocations. What you are seeing here is probably something different, though. NetworKit allocates a separate `std::vector` (i.e., a dynamic array for those not familiar with C++) for every node. This allows dynamically adding and removing edges. This comes at a cost, though. Each of these vectors has 24 bytes overhead for storing the allocated size, the actual size and the pointer to the data. This means for every node both for in- and out-edges and weights you get 24 * 4 = 96 bytes overhead. Note that edge ids are only allocated after you called `indexEdges()` on the graph and thus do not need to be considered here. In addition to that you have the 2 * 8 = 16 bytes for storing the incoming or outgoing edge and their weight of each of these nodes. This times 20 million gives 2.08 GiB according to my calculation. Each non-empty vector also needs a memory allocation. According to [this Stack Overflow post](https://stackoverflow.com/questions/15881170/is-there-a-memory-overhead-associated-with-heap-memory-allocations-eg-markers-i), each memory allocation has 32-64 bytes overhead. As two vectors per node are non-empty and thus need a memory allocation, we get an extra 64-128 bytes per node, or another 1.19-2.38 GiB. At least the latter would explain your observed memory usage. Further, memory allocations might be aligned to 16 bytes, so edges might actually cost 2*16 bytes per node.

So in other words, degree-one-nodes in directed graphs are basically the worst case in terms of memory overhead per edge. NetworKit expects significantly larger node degrees and is thus not optimized for such graphs.

Concerning plain text formats: In my experience, METIS graphs are much faster to read and result in less fragmented memory allocations (and thus less memory usage in total) than edge lists because all edges for a node are added at the same time. They only support undirected graphs, though. I think it would be good to have a similar (maybe even the same) format for directed graphs.

**Comment by AndrewHannigan** [comment]
@michitux Very valuable insight for me, thank you for the explanation.  Wanted to share some additional calculations I think might be of interest. Let's assume an unweighted fully connected directed graph for simplicity so that every node has non-empty in and out edge lists with n nodes and m edges, meaning we've got two allocated and non-empty std::vectors for each node.

Node Array = n * 8 bytes
Incoming Edge List Overhead = n * (24 bytes for std::vector internals + 64 bytes OS memory allocation of the vector)
Outgoing Edge List Overhead= n * (24 bytes for std::vector internals + 64 bytes OS memory allocation of the vector)
Actual Edge List Data = 2m * 8 bytes

This works out to:
Memory consumption from std::vector overhead = 176n
Overall memory consumption = 184n + 16m

So the ratio of std::vector overhead to overall memory consumption is 176n / (184n + 16m).  If we set that to 50%, we get m/n = 10.5. 
 So for graphs with avg degree > 10, the graph data dominates memory.  And for graphs with avg degree < 10, std::vector overhead dominates memory.  I'm ignoring the discrete size classes of the vectors and memory allocation alignment, so in reality the crossover point is actually higher than 10 but it's a helpful lower bound.  So degree-one nodes are certainly a worst case, and importantly you really don't overcome the std::vector overhead until your graph has avg degree of at least 10.  

All this to say, I do want to make the case for addressing this overhead for the purposes of improving networkit's peak memory performance for low degree graphs.  My own motivating example for this is cryptocurrency/blockchain analytics, where the avg number of transactions per address is roughly 2, thus a sparse yet massive graph.  

One possible solution is to create a `StaticGraph` structure that doesn't allow modifications after the graph is instantiated.  The LightGraph project for Julia has something like that.  That would give a lot of options in terms of ways to seriously trim down overhead.  The primary modification I would foresee is 
* replacing the n individual edges-out std::vectors with one simple array of length m to store edge-out lists for all nodes.
* defining a lookup array of length n that points to each node's starting address in the edge-out array
* defining a lookup array of length n that indicates the out degree of each node, so we know how far to read in the edge-out list for that node

That is just my guess, and apologies in advance as I'm sure I'm oversimplifying.  But would be curious to know whether others are also motivated to address the memory overhead for sparse graphs and if there is something like this proposed solution on the roadmap?

**Reply by avdgrinten** [reply]
What you are describing is essentially a CSR matrix. We do not want to turn the `Graph` class into a data structure that only supports static graphs, but there are also other ways to optimize the memory usage of low-degree graphs (such as custom dynamic array data structures in place of `std::vector`). We already have somebody working on it, but some engineering and experimentation needs to be done until we can merge it.

**Reply by AndrewHannigan** [reply]
@avdgrinten Yes good point, basically CSR matrix without the COL_INDEX since adjacency lists don't have an ordering inherent to the graph.  Is there an open issue/discussion you could point me towards on the custom dynamic array?  I found this but it looks like public discussion petered out at the start of last year: https://github.com/networkit/networkit/discussions/691

Very interested in following the progress and assisting on that feature however possible, whether through contribution or through sponsorship via [TRM Labs](https://blog.trmlabs.com/post/north-korea-cyber-attack).
